### PR TITLE
Typeclasses for decoding EJson

### PR DESCRIPTION
**Pull request by wemrysi**
Adds `DecodeEJson` and `DecodeEJsonK` typeclasses for decoding values from `EJson` along with instances for types that had already defined an `EncodeEJson` instance.

### CODE REVIEW

File: frontend/src/main/scala/quasar/sst/TypeStat.scala [hunk 1]
```diff
@@ -217,64 +302,83 @@ sealed abstract class TypeStatInstances {
...
-          "true"  -> t.asEJson[J],
-          "false" -> f.asEJson[J])
-      case Byte(c, mn, mx)          => minmax("byte", c, mn, mx)
-      case Char(c, mn, mx)          => minmax("char", c, mn, mx)
-      case Int(s, mn, mx)           => dist("integer", s, mn, mx)
-      case Dec(s, mn, mx)           => dist("decimal", s, mn, mx)
-      case Count(c)                 => kmap("count", "count" -> c.asEJson[J])
-
-      case Coll(c, mnl, mxl)        =>
+          "boolean"
```
Comment (rintcius) [inline frontend/src/main/scala/quasar/sst/TypeStat.scala]: maybe good to also extract these as constants & deduplicate them?

File: frontend/src/main/scala/quasar/tpe/TypeF.scala [hunk 2]
```diff
@@ -356,51 +357,91 @@ private[quasar] sealed abstract class TypeFInstances {
       )
     }
 
+  implicit def decodeEJsonK[A](implicit A0: DecodeEJson[A], A1: Order[A]): DecodeEJsonK[TypeF[A, ?]] =
+    new DecodeEJsonK[TypeF[A, ?]] {
+      def decodeK[J](implicit JC: Corecursive.Aux[J, EJson], JR: Recursive.Aux[J, EJson]): CoalgebraM[Decoded, TypeF[A, ?], J] =
+        j => j.decodedKeyS[String](TypeKey) flatMap {
+          case "bottom" =>
```
Comment (rintcius) [inline frontend/src/main/scala/quasar/tpe/TypeF.scala]: also extract these as constants?

**Comment by wemrysi** [comment]
@rintcius Thanks for the feedback, I've pushed a couple commits that address your comments.

**Comment by rintcius** [comment]
:+1:

**Pull request merged**